### ISSUE TITLE
Counter increments even if document validation fails

### ISSUE THREAD

**Issue by DrNightmare** [nodkz/mongoose-plugin-autoinc#10]
On trying to save a document that fails the Mongoose validation specified in the schema, the counter for the corresponding collection still gets incremented but the document does not get saved (as expected). This results in the auto incremented ids not being continuous.

Could we add a flag that checks if the validation fails for a document and not increment the counter in that case? I had a look at the [pre hook](https://github.com/nodkz/mongoose-plugin-autoinc/blob/d2f48099bd5c79bcafc19f15e2f60defe8849c73/src/index.js#L321) on 'validate' and this could be a potential fix:

Replace:
`if ((doc.isNew && !alreadyGetId) || settings.migrate) {`

with:
`if ((doc.isNew && !alreadyGetId && doc.validateSync === undefined) || settings.migrate) {`

I'm not sure if this is the best way to achieve this, but this could be a start.

**Comment by nodkz** [comment]
Autoincrement should not call **validation**, cause it may break somebody logic and produce unnecessary overhead and unobvious behavior.

You may disable auto increment by:
```js
doc.__maiRanOnce = true;
```

But it is not clear and internal property. So let to expose this via adding a new public method for the document (like it did for [another stuff](https://github.com/nodkz/mongoose-plugin-autoinc/blob/d2f48099bd5c79bcafc19f15e2f60defe8849c73/src/index.js#L300-L303)):
```js
  schema.method('disableAutoIncrement', function() {
    const doc: MongooseDocument = this;
    doc.__maiRanOnce = true;
  });
```

PR are welcome. And please don't forget to add tests.
Thanks.

**Comment by DrNightmare** [comment]
Good point on not calling validation from autoincrement. I however fail to see how I would disable the auto increment selectively for documents that fail validation, without calling the validation function. 

Since the schema itself runs validate and then save, maybe we should just do the autoincrement on the post hook for validate?

This would be a much cleaner fix. We could just replace the [pre hook](https://github.com/nodkz/mongoose-plugin-autoinc/blob/d2f48099bd5c79bcafc19f15e2f60defe8849c73/src/index.js#L313) on 'validate'

with a post hook:
`schema.post('validate', function(next: Function) {`

Thoughts?

**Comment by sp3c1** [comment]
I dont know, to tell the truth It would be expected behaviour for me.

**Comment by AlexTemina** [comment]
Interested in this. Same here.

Thanks for solving the issue.
